src/coreclr/vm/i386/asmhelpers.asm

```diff
@@ -1142,60 +1142,33 @@ _VarargPInvokeStub@0 endp
 ; Invoked for marshaling-required unmanaged CALLI calls as a stub.
 ; EAX       - the unmanaged target
 ; ECX, EDX  - arguments
-; [ESP + 4] - the VASigCookie
+; EBX       - the VASigCookie
 ;
 _GenericPInvokeCalliHelper@0 proc public
     ; save the target
     push    eax
 
     ; EAX <- VASigCookie
-    mov     eax, [esp + 8]           ; skip target and retaddr
-
-    mov     eax, [eax + VASigCookie__StubOffset]
+    mov     eax, [ebx + VASigCookie__StubOffset]
     test    eax, eax
 
     jz      GoCallCalliWorker
     ; ---------------------------------------
 
-    push    eax
-
-    ; stack layout at this point:
-    ;
-    ; |         ...          |
-    ; |   stack arguments    | ESP + 16
-    ; +----------------------+
-    ; |     VASigCookie*     | ESP + 12
-    ; +----------------------+
-    ; |    return address    | ESP + 8
-    ; +----------------------+
-    ; | CALLI target address | ESP + 4
-    ; +----------------------+
-    ; |   stub entry point   | ESP + 0
-    ; ------------------------
-
-    ; remove VASigCookie from the stack
-    mov     eax, [esp + 8]
-    mov     [esp + 12], eax
-
-    ; move stub entry point below the RA
-    mov     eax, [esp]
-    mov     [esp + 8], eax
-
-    ; load EAX with the target address
-    pop     eax
-    pop     eax
-
     ; stack layout at this point:
     ;
     ; |         ...          |
     ; |   stack arguments    | ESP + 8
     ; +----------------------+
     ; |    return address    | ESP + 4
     ; +----------------------+
-    ; |   stub entry point   | ESP + 0
-    ; ------------------------
+    ; | CALLI target address | ESP + 0
+    ; +----------------------+
+
+    ; load EAX with the target address, store on stack stub entry point
+    xchg    eax, [esp]
 
-    ; CALLI target address is in EAX
+    ; call stub
     ret
```

Comment:
@jakobbotsch Should I change this `ret` instruction to `jmp` also?

Reply:
I think it's not important for these helpers for the reasons @jkotas mentioned. The write barriers are much hotter (though not sure about the callable version).
But if it makes the stubs simpler then go for it.

Reply:
It would be nice to change this to `jmp` also. It should be easy to do now that there is an extra register to use.

Reply:
Thanks, done.